enhancements/etcd/etcd-encryption-for-separate-oauth-apis.md

```diff
@@ -0,0 +1,149 @@
+It holds an array of keys that are used to encrypt resources. 
+
+OpenShift maintains a single `EncryptionConfiguration` resource that is created and maintained by a set of controllers called the encryption controllers in `openshift-config-managed` namespace for each API server that needs encryption.
+For example `OAS-O` creates `encryption-config-openshift-apiserver` secret in that namespace. At a later stage it is copied to `openshift-apiserver` namespace, revisioned and finally makes its way to the API server.
+
+In order to make two API servers (`openshift-apiserver` and `oauth-apiserver`) use the same `encryption-config` we are going to create a copy for the new API and let it be managed by `OAS-O` for one release `n` to support downgrades.
+On the next release `n+1` we will copy over the keys to avoid creating new ones. From that point on `authentication-operator` will maintain its own config.
+
+1. Create a new controller in `OAS-O` that will create and annotate `encryption-config-openshift-oauth` secret in `openshift-config-managed` namespace. It also must keep it in sync with `encryption-config-openshift-apiserver`.
+ The annotation will prevent `CAO` from managing the newly created `encryption-config`.
+2. Prepare `authentication-operator` to manage `encryption-config-openshift-oauth` but only if it doesn't have an annotation.
+   - make use of `encryption.NewControllers` but don't start it if the annotation is present
+3. Update `CAO` to revision and plug `encryption-config` for its operand.
+4. Create a new deployer (`statemachine.Deployer`) called `UnionRevisionLabelPodDeployer` that will manage multiple `RevisionLabelPodDeployer`.
+5. Implement `DeployedEncryptionConfigSecret() (secret *corev1.Secret, converged bool, err error)` function that:
+   - calls `DeployedEncryptionConfigSecret` for all components
+   - returns "failure" if any component has reported an error (`err !=nil`) or hasn't yet converged (`converged == false`)
+   - returns "failure" if the `secret` resource differs among components. In oder to compare the `secret`:
+     - use `encryptionconfig.FromSecret` function that returns `EncryptionConfiguration`
+     - use `reflect.DeepEqual` function to compare `EncryptionConfigurration.Resources` for example `reflect.DeepEqual(openshiftAPIServerEncryptionCfg.Resources, oauthAPIServerEncryptionCfg.Resources)`
+   - returns "success" otherwise
+6. Change `NewRevisionLabelPodDeployer` to conditionally keep the secrets in synchornization. 
+7. Update `OAS-O` to use `UnionRevisionLabelPodDeployer` passing two `RevisionLabelPodDeployer`. The first one for `openshift-apiserver` (already existing) and the second one for `oauth-apiserver`.
+   - it will report whether all instances of `oauth-apiserver` converged to the same revision
+   - make sure that `oauth-apiserver` deployer won't synchronize the secrets
+
```

Comment:
do we have to think about the conditions, especially the `Encrypted` condition?

Reply:
I don't think so. I think that condition is set based on the current state (reported by the deployer) and annotations assigned to `secrets` - both conditions will be met.